### DISCUSSION TITLE
Excavation with a huge number of blocks

### DISCUSSION THREAD

**Discussion by jmeier** [idaholab/moose#26505] [Q&A Modules: Solid mechanics] [answered]
Dear Moose-Community,

I have seen that the topic of simulating excavation stages has already been discussed several times in the [Discussions](https://github.com/idaholab/moose/discussions?discussions_q=excavation) and that there is also a very nice [demo-model](https://mooseframework.inl.gov/modules/porous_flow/coal_mining.html) for this. However, the examples found can either describe the excavation front with an x-y-z function or have a small number of blocks.

I would like to create a 3D model in which a large number of blocks are successively excavated and backfilled over time. The blocks (more than 5000) are arranged in such a way that no "excavation front" can be reasonably described mathematically as an equation. Excavation and refilling must therefore be defined per block over time.

A few pieces of the puzzle that I have found so far:
-  [CoupledVarThresholdElementSubdomainModifier](https://mooseframework.inl.gov/source/userobjects/CoupledVarThresholdElementSubdomainModifier.html): Using this userobject, one can "activate"/"deactivate" blocks over time. 
   - Plus: Using the "block" parameter i can restrict this userobject to a singe block. Thus, function to be given in "coupled_var" needs only depend on time.
   - Minus: Moving a block to another subdomain takes place abruptly.
   - Question 1: If I use this user object, I have to create thousands of instances of the CoupledVarThresholdElementSubdomainModifier (e.g. via script). Given this amount of UserObjects, do I have to expect performance or memory problems in Moose? 
   - Question 2: In case there are performance or memory problems for a huge number of  CoupledVarThresholdElementSubdomainModifier - instances, is there a way to reduce the number of userobjects e.g. using a CSV-file? I did not find a way to make the function to be given in "coupled_var" to be dependent on the blocks name and using the coordinates seems messy.
		
- Pre-Factoring Material Properies: In order to avoid (or prepare) the sudden removal of blocks, the stiffness can be reduced in a time-dependent manner using "elasticity_tensor_prefactor", for example (density etc. likewise; e.g. #23677)
    - Plus: Numerical stability.
    - Question 3: Because the blocks are to be excavated one by one, I would have to create a *huge* number of materials (worst case: one material set per block). This I can do easily by script. But: Given this amount of materials, do I have to expect performance or memory problems in Moose?
    - Question 4: In case there are performance or memory problems for a huge number of  materials, is there a way to reduce the number of materials e.g. using a CSV-file? I did not find a way to make the function to be given in "elasticity_tensor_prefactor" to be dependent on the blocks name.

I hope to be able to combine both approaches above: First "fade out" a block by factorising the material parameters and then remove the block completely.

Kind regards, Jörg

**Comment by GiudGiud** [comment]
Hello

A few comments. Ultimately, whether the performance is acceptable for you or not depends on your machine and your patience so that's just something you should find out. Objects are replicated across parallel domains though so there will be a penalty from using thousands of them, especially in memory. How big is the mesh? How big are each node if using HPC?

But 
> [CoupledVarThresholdElementSubdomainModifier](https://mooseframework.inl.gov/source/userobjects/CoupledVarThresholdElementSubdomainModifier.html):
> Plus: Using the "block" parameter i can restrict this userobject to a singe block. Thus, function to be given in "coupled_var" needs only depend on time.

you could just use one of these objects and not-block restricti it. I think you should aim to capture the block-dependent behavior in the threshold instead. 

> Question 3: Because the blocks are to be excavated one by one, I would have to create a huge number of materials

same idea. Just make a block restricted tensor prefactor instead if you do not want to have thousands of materials. (what you say in Q4)
Thousands of materials will scale well CPU-wise, they are not all active at the same time on every element. 

> I did not find a way to make the function to be given in "elasticity_tensor_prefactor" to be dependent on the blocks name.

this has not been needed before.  But the way to do this will be to use a material property OR a variable OR use the functor interfaces (most flexible) for this parameter. The material will need a slight modification to do that. 

I would recommend starting small. One or two blocks at most. 

Best,
Guillaume

**Reply by jmeier** [reply]
Dear Guillaume,

Thank you for your reply and your comments.

I don't have an HPC available. That's why I want to be careful with the resources. The model will probably consist of 300'000 to 400'000 TET10 elements.

> I think you should aim to capture the block-dependent behaviour in the threshold instead.

That was the aim of my Q2 and Q4: I'm not yet clear on the best way to do this. I have seen that [PropertyReadFile](https://mooseframework.inl.gov/source/userobjects/PropertyReadFile.html) has a parameter "read_type", which I can set to "block". 
- Under [PropertyReadFile](https://mooseframework.inl.gov/source/userobjects/PropertyReadFile.html) it is specified that the order of the blocks in the CSV must match that in the Moose (and the IDs must start with 1). Alternatively, can I enter the block name in the first column of the CSV?
- Therefore: The rows of the CSV correspond to the blocks. So I would like to define when which block is on/off in the columns. 
- How do I map the data in the columns of the CSV to my model time? In other words: How do I select a column of the CVS depending on the model time "t"?

Kind regards,
Jörg

**Reply by GiudGiud** [reply]
> Alternatively, can I enter the block name in the first column of the CSV?

that has not been implemented but you could do it if you want to. 

> How do I map the data in the columns of the CSV to my model time?

you currently cannot. We have not added a time column. That's also something that could be added
You can specify multiple CSV files and change every time step / every time the PropertyReadFile is executed (can be more or less frequent than time steps)

**Reply by jmeier** [reply]
Dear Guillaume,

>>Alternatively, can I enter the block name in the first column of the CSV?

>that has not been implemented but you could do it if you want to.

Unfortunately, my C++ skills are are not sufficient. However, this functionality would certainly be something that would make the reading of block-related data much more robust in Moose. Perhaps something for the wish-list?
  

>We have not added a time column. That's also something that could be added
>You can specify multiple CSV files and change every time step / every time the PropertyReadFile is executed (can be more or less frequent than time steps)

I've seen the discussion going on in #26294. For my model I expect a lot of automatically generated time steps. So providing one CSV per time step is no option (due to fact I do not know the time steps chosen by Moose). So, I presume for me it would be favorable to use the approach you describe as:

> So if you used an auxkernel with let's say 10 CSV files for ten time steps you need:
> - 10 [PropertyReadFile](https://mooseframework.inl.gov/source/userobjects/PropertyReadFile.html) to read each file
> - 10 functions [PiecewiseConstantFromCSV](https://mooseframework.inl.gov/source/functions/PiecewiseConstantFromCSV.html), one for each UO
> - 10 auxkernels [FunctionAux](https://mooseframework.inl.gov/source/auxkernels/FunctionAux.html) that set the SAME auxvariable damage using a different function
> - 10 Controls objects that turn on/off the auxkernels on each time step

... and then use the auxvariable inside the [CoupledVarThresholdElementSubdomainModifier](https://mooseframework.inl.gov/source/userobjects/CoupledVarThresholdElementSubdomainModifier.html)? If this is correct, what type of AuxVariable do I have to define to maintain "read_type='block'" given in PropertyReadFile and PiecewiseConstantFromCSV?

When pre-factoring material properies, I presume I need 20 auxkernels with this approach (10 writing one auxvariable, and another one writing another auxvariable). When pre-factoring I use these two auxvariables and interpolate? 

Kind regards,
Jörg

**Reply by GiudGiud** [reply]
> So, I presume for me it would be favorable to use the approach you describe as:

no this is an old approach. 
You can do this approach with many Controls object (one for every file change you want) and specifying all the CSV files in a single PropertyReadFile, with a single PiecewiseConstrantFromCSV etc.

**Reply by GiudGiud** [reply]
> When pre-factoring I use these two auxvariables and interpolate?

this is not a bad idea actually if you want a smoothly varying pre-factor. I still would only use 2 auxvariables, so 2 functions, 2 FunctionAux and 2 PropertyReadFile (one for each side of the interpolation)

**Reply by jmeier** [reply]
Dear @GiudGiud,

> You can do this approach with many Controls object (one for every file change you want) and specifying all the CSV files in a single PropertyReadFile, with a single PiecewiseConstrantFromCSV etc.

Having only one [PropertyReadFile](https://mooseframework.inl.gov/source/userobjects/PropertyReadFile.html) with a single [PiecewiseConstrantFromCSV](https://mooseframework.inl.gov/source/functions/PiecewiseConstantFromCSV.html) sound for sure much more efficient. Unfortunately, it is not yet clear to me how to map the sequence of CSV files to specific model times.

My [Executioner] block looks like:
```
# Executioner defintion
[Executioner]
  type = Transient
 
  #set solver
  solve_type = 'PJFNK'
 
  # options for petsc
  petsc_options = '-snes_converged_reason'
  petsc_options_iname = '-pc_type -pc_factor_mat_solver_package'
  petsc_options_value = ' lu       mumps'
 
  ...
 
  # timing options
  start_time = 0.0
  end_time   = 1.25
  dt         = 0.25
  dtmin      = 0.0001
[]
```

In the [Executioner]-block above I give a `dt` - but in most of my models the 'real' timesteps are smaller and more numerous (as allowed by `dtmin`). Therefore my question regarding the mapping of the CSV-files to the timesteps. 

Kind regards, Jörg

**Reply by GiudGiud** [reply]
Do you know the exact times you need to change CSV files? 

If so, write a ParsedFunction that is equal to 1 at these times and 0 elsewhere. 

Then use the
conditionalFunctionEnableControl 
To enable the reader at those times. 


My other idea is to write a new Control object that activates only at specified times. I m busy with a training right now but I can make that next week maybe.

**Reply by jmeier** [reply]
Dear @GiudGiud,

After the distraction of #26664 I removed all `${units ... }` from my test project for now and tried to follow your recommendations. Please find the current state of my test project [this](https://github.com/grunerjmeier/moose-blockystages) repository.

The model consists of 7 blocks. 6 of them are to be activated/deactivated at pre-known times. The times are stored in [stages.csv](https://github.com/grunerjmeier/moose-blockystages/blob/main/stages.csv). The corresponding activation states in [stages.block_activation_t0000.csv](https://github.com/grunerjmeier/moose-blockystages/blob/main/stages.block_activation_t0000.csv) to [stages.block_activation_t0003.csv](https://github.com/grunerjmeier/moose-blockystages/blob/main/stages.block_activation_t0003.csv). 

I can run the model and Moose is calculating something. But...

- Moose creates not one .e-file, but a set of .e-s### files.
- In Paraview, only one block is available/visible (the big one).
- My aux variable `BlockActivation_StateAuxVariable` is always shown to be zero by Paraview (despite the fact it is set to 1 at distinct times in `BlockActivation_AdvanceStatesFileReaderFunction`

Obviously, I'm doing something wrong. Could you please have a look?

Kind regards,
Jörg

**Reply by GiudGiud** [reply]
> Moose creates not one .e-file, but a set of .e-s### files.

this is because the mesh is changing. Are you using adaptivity for example?
It s a limitation of the exodus format.

> In Paraview, only one block is available/visible (the big one).

In the left panel you only see on block? And in the moose simulation there are more? (for example in the header of the simulation it reports the number of blocks)

> My aux variable BlockActivation_StateAuxVariable is always shown to be zero by Paraview (despite the fact it is set to 1 at distinct times in BlockActivation_AdvanceStatesFileReaderFunction

Do you see the auxiliary kernel running using `Debug/show_execution_order = ALWAYS` ?

**Reply by jmeier** [reply]
Dear @GiudGiud,

> this is because the mesh is changing. Are you using adaptivity for example?

I do not use adaptivity (at least not deliberately). But I use a `CoupledVarThresholdElementSubdomainModifier` to move blocks between subdomains:
```
[UserObjects]
  [BlockActivation]
    type = CoupledVarThresholdElementSubdomainModifier
    coupled_var = 'BlockActivation_StateAuxVariable'
    criterion_type = BELOW
    threshold = 0
    subdomain_id = 1
    complement_subdomain_id = 2
    execute_on = 'INITIAL TIMESTEP_BEGIN'
  []
[]
```

> In the left panel you only see on block? And in the moose simulation there are more? (for example in the header of the simulation it reports the number of blocks)

Yes, only one "block" in Paraview. Moose reports "Num of subdomains: 7" (`Mesh/File/show_info=true` also reports 7 "subdomains").
By the way: Is there a mixup between the terms "subdomain" and "block"? This makes me a bit nervous due to the fact the `CoupledVarThresholdElementSubdomainModifier` is moving between `subdomain_id` and `complement_subdomain_id `. Is this my mistake?
```
Mesh:
  Parallel Type:           replicated
  Mesh Dimension:          3
  Spatial Dimension:       3
  Nodes:                   5207
  Elems:                   3325
  Num Subdomains:          7
```
If I run the model without any of the stuff connected with the `CoupledVarThresholdElementSubdomainModifier`, I see the 7 blocks in Paraview just fine.

> Do you see the auxiliary kernel running using `Debug/show_execution_order = ALWAYS` ?

In the Debug I see before the first step that Moose is handling 7 blocks. This is the message for the last block:
```
[DBG] Only user objects active on local element/sides are executed
[DBG] Ordering of User Objects on block 7
[DBG] Executing ElementUserObject on INITIAL
[DBG] Order of execution:
[DBG] BlockActivation
```
For all other timesteps Moose is only talking about "block 2". This is the message for time step 3:

```
Time Step 3, time = 0.6, dt = 0.2
[DBG] Computing elemental user objects on TIMESTEP_BEGIN
[DBG] Ordering of User Objects on block 2
[DBG] Executing ElementUserObject on TIMESTEP_BEGIN
[DBG] Order of execution:
[DBG] BlockActivation

[DBG] Only user objects active on local element/sides are executed
[DBG] Executing auxiliary kernels on elements on LINEAR
[DBG] Ordering of AuxKernels on block 2
[DBG] BlockActivation_StateAux
      Still Computing Residual..[DBG] Beginning elemental loop to compute Residual on LINEAR
[DBG] Ordering of Residual Objects on block Box2 (2)
[DBG] Ordering of kernels:
[DBG] TM_all0 TM_all1 TM_all2 Gravity
[DBG] Ordering of Residual Objects on boundary BoundaryZMax (18)
[DBG] Ordering of integrated boundary conditions:
[DBG] BoundaryZMax_Pressure

      Finished Computing Residual                                                        [ 22.65 s] [  204 MB]
    Finished Computing Initial Residual                                                  [ 22.65 s] [  204 MB]
[DBG] Executing auxiliary kernels on elements on LINEAR
[DBG] Ordering of AuxKernels on block 2
[DBG] BlockActivation_StateAux
    Still Computing Residual..[DBG] Beginning elemental loop to compute Residual on LINEAR
[DBG] Ordering of Residual Objects on block Box2 (2)
[DBG] Ordering of kernels:
[DBG] TM_all0 TM_all1 TM_all2 Gravity
[DBG] Ordering of Residual Objects on boundary BoundaryZMax (18)
[DBG] Ordering of integrated boundary conditions:
[DBG] BoundaryZMax_Pressure

    Finished Computing Residual                                                          [ 23.51 s] [  204 MB]
 0 Nonlinear |R| = 5.134485e-06
 Solve Converged!
  Finished Solving                                                                       [ 46.17 s] [  204 MB]
[DBG] Initializing, executing & finalizing general UO 'BlockActivation_StatesCSV' on TIMESTEP_END
[DBG] Executing auxiliary kernels on elements on TIMESTEP_END
[DBG] Ordering of AuxKernels on block 2
[DBG] stress_xx_all stress_xy_all stress_xz_all stress_yy_all stress_yz_all stress_zz_all BlockActivation_StateAux

Still Executing..
Outlier Variable Residual Norms:
  disp_z: 4.984422e-06
```

**Reply by GiudGiud** [reply]
> I do not use adaptivity (at least not deliberately). But I use a CoupledVarThresholdElementSubdomainModifier to move blocks between subdomains: 

That will do it too. Another limitation of the output format

> Yes, only one "block" in Paraview. Moose reports "Num of subdomains: 7" (Mesh/File/show_info=true also reports 7 "subdomains").

A subdomain is the same as a block. The block name comes from paraview iirc and has permeated into our naming. It used to be subdomain, which is the libmesh name. 

> If I run the model without any of the stuff connected with the CoupledVarThresholdElementSubdomainModifier, I see the 7 blocks in Paraview just fine.

Interesting.
When you do use the CoupledVaretcModifier, do you see the elements still but with the wrong subdomain label?

Given the console log I am tempted to think all elements are now in block 2. I think this might be due to the CoupledVaretcmodifier.
I think setting the `complement_subdomain_id` is NOT what you want to do. It moveds everything that is not meant to be in block 1 (due to missing the threshold) into block 2.

**Reply by jmeier** [reply]
Dear @GiudGiud,

> That will do it too. Another limitation of the output format

Thank you for the clarification. Am I right in assuming that Paraview loads all e-files of this 'batch' as a whole and automatically switches forth and back between the files?

> When you do use the CoupledVaretcModifier, do you see the elements still but with the wrong subdomain label?
Yes, that is exactly the case.

By the way: I used `complement_subdomain_id` to be able to re-activate the blocks in a later model stage. As I understood what is written [here](https://mooseframework.inl.gov/source/userobjects/CoupledVarThresholdElementSubdomainModifier.html), one has to do this.

I experimented a bit more:

- If I set `complement_subdomain_id = 100` in `CoupledVarThresholdElementSubdomainModifier`, Moose gives a message rading "Requested subdomain 100 does not exist."
- When I comment out `complement_subdomain_id`, I get again 7 blocks. But the blocks are _not_ getting activated/deactiated. Please see image 1 below. 

Image 2 below shows the model when I use `complement_subdomain_id = 2`: All my elements are there - but only one block.

This makes me wondering: Do I have the right command with `CoupledVarThresholdElementSubdomainModifier`? My goal is to deactivate (and later sometime to re-activate) blocks in my simulation to model excavation.

image 1: the 7 blocks color coded
![grafik](https://github.com/idaholab/moose/assets/57180520/fae6e4ba-a8de-4e92-a540-7233df9a50b1)

image 2: all elements present - but no blocks using`complement_subdomain_id = 2`
![grafik](https://github.com/idaholab/moose/assets/57180520/2f6cefe4-cddc-4cba-a5c8-9a3a13edf5ec)

**Reply by jmeier** [reply]
I just found #20836 offering some hints on how to move elements to other subdomains. Let me go through this first...

**Reply by jmeier** [reply]
Hi @GiudGiud, 

Going through #20836 left me with a question I'd like to check with you: Moving elements via `CoupledVarThresholdElementSubdomainModifier` to subdomains without physics which must not be empty adds some complexity to a model definition. On the other hand, if I understand your [comment](https://github.com/idaholab/moose/discussions/20836#discussioncomment-2616231) correctly, if I restrict the physics from a block, it will ignored by the solver. Due to the fact I always plan to remove full blocks from a model (and later activating them again as full block), is there a way to just remove physics from a block without moving them to other subdomains? Would this be an alternative (easier) way to model "excavation" and "refill"?

Thank you for your time and patience!

Kind regards, Jörg

**Reply by GiudGiud** [reply, marked as answer]
> Thank you for the clarification. Am I right in assuming that Paraview loads all e-files of this 'batch' as a whole and automatically switches forth and back between the files?

no it does not you need to select all of them when loading the file (to load multiple files at once) iirc.

> By the way: I used complement_subdomain_id to be able to re-activate the blocks in a later model stage. As I understood what is written [here](https://mooseframework.inl.gov/source/userobjects/CoupledVarThresholdElementSubdomainModifier.html), one has to do this.

From the documentation

> If the CoupledVarThresholdElementSubdomainModifier is applied onto the entire domain, and the parameter complement_subdomain_id is set to 2, then the subdomain ID of all elements outside the circle will be set to 2:

so yes if you want it to be reversible, you need to use a complement subdomain id. But I think you could also simply use a second SubdomainModifier UO to model the reverse transformation. 

> If I set complement_subdomain_id = 100 in CoupledVarThresholdElementSubdomainModifier, Moose gives a message rading "Requested subdomain 100 does not exist."

you can pre-declare this subdomain in the mesh block with the `add_subdomain_ids` parameter. See Mesh doco page for exact parameter spelling

>  Due to the fact I always plan to remove full blocks from a model (and later activating them again as full block), is there a way to just remove physics from a block without moving them to other subdomains?

Not that i know. The problem is that if there are no equations for the degrees of freedom (DOFs) living on that subdomain, then the system is ill-posed and it most likely will not converge.
You could have a dummy equation like just the time derivative = 0. To do that you could use the Control system (in the Controls block) with the ConditionalEnableControl to deactivate some kernels based on a condition

**Reply by jmeier** [reply]
Dear @GiudGiud,

Thanks a lot for your valued input. I have tried to summarise what I have learned in a short show & tell: #26941

Kind regards,
Jörg

**Comment by Traiwit** [comment]
Hi, by 'excavation' do you want to actually remove the blocks from the mesh, or you are okay with changing the blocks property to be really weak (lower the youngs_modulus or something like that)

if so, you can have the state map/flag material for each element (via PropertyReadFile), and modify the material property of each element (for example youngs_modulus) based on the state

PropertyReadFile can read .csv file and give the value for each element (1 to n_th element == 1 to n_th row of the file)

I did this with PorousFlow permeability, but I think it's the same in Tensormechanics

```
  // calculate perm tensor
  if (_flag[_qp] == 1){
    // state==1: this is for excavated domains, set permeability to 1e-18 = very low

      _perm[_qp](0) =  _perm_excav; 
      _perm[_qp](1) =  _perm_excav; 
      _perm[_qp](2) =  _perm_excav; 

  }
  else if (_flag[_qp] == 2) {
    // state==2: this is for caved domains, set permeability to 1e-14 (kw=1e-5)
    _perm[_qp](0) =  _perm_cave; 
    _perm[_qp](1) =  _perm_cave; 
    _perm[_qp](2) =  _perm_cave; 

  }
  else if (_flag[_qp] == 3 ) {
  
    _perm[_qp](0) =  _perm_backfill; 
    _perm[_qp](1) =  _perm_backfill; 
    _perm[_qp](2) =  _perm_backfill; 
  }
  else {
    // any other state (including state==0): this is for intact domains, set permeability to exponential function of damage
```

my C++ is very bad but ChatGPT helps a lot. Personally I find modifying mesh during the simulation is very difficult.

Hope this helps, I also do mining applications like excavation/backfill/cave but for hydro - with Abaqus mesh, I hope to develop a system for Tensormechanics soon (still waiting for Cohesive elements so I can properly deal with faults).

**Reply by jmeier** [reply]
Dear @Traiwit, 

Thank you for your message. 

In the [coal mining model](https://mooseframework.inl.gov/modules/porous_flow/coal_mining.html) actually the "softening" you are proposing is done. But thanks to the `elasticity_tensor_prefactor`, no C++ code is needed in tensor mechanics (in combination of reducing the weight of the blocks affected).

In view of experiences with other FE codes / FE models, I am very interested in really removing the blocks (and possibly reinserting them later). Depending on the model, the approach with the `elasticity_tensor_prefactor` is also quite tricky: For most tasks, the `elasticity_tensor_prefactor` alone is not sufficient. The material also often has to become purely linear-elastic. And lighter (see also the coal mining model). And more permeable. And... And yet the edges of these blocks would not be stress-free. Compressive stresses - if not too large - are often not a problem. Tensile stresses, on the other hand, are a major problem if the "normal" blocks next to them have no or very low tensile strength (like many soft soils like sand or clay). In addition, other FE codes quickly encountered problems with convergence if the stiffness differences and deformations of the "soft ghost blocks" were too great.

Therefore, the idea was to first soften the blocks to be excavated (using `elasticity_tensor_prefactor` and density and ...) and then to really remove them from the model.

Kind regards,
Jörg
